### PR TITLE
`Programming exercises`: Upgrade default java dependencies

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context
The programming templates should keep their dependencies up to date.

### Description
Updated dependencies and plugins to their newest version.

We should stay consistent with the versions present in the docker image. Relevant PR: https://github.com/ls1intum/artemis-maven-docker/pull/135

### Steps for Testing
Create a new programming exercise, check that running the template and solution still works, also check that static code analysis feedback is visible.

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
  - Updated Gradle distribution URL to version 8.7 across various templates.
  - Added new PMD rules for code analysis and removed outdated ones.
  - Updated Apache Commons library to version 3.14.0 in build configurations.
- **Bug Fixes**
  - Improved clarity and consistency of error messages related to `JAVA_HOME` settings in batch files.
- **Documentation**
  - Enhanced error message outputs to stderr in various script files.
- **Refactor**
  - Adjusted handling of JVM options and shell script fragments for better script performance and reliability.

### CODE REVIEW

File: src/main/resources/templates/java/plain_gradle/exercise/gradlew.file [hunk 1]
```diff
@@ -202,11 +202,11 @@ fi
 # Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
 DEFAULT_JVM_OPTS='"-Xmx64m" "-Xms64m"'
 
-# Collect all arguments for the java command;
-#   * $DEFAULT_JVM_OPTS, $JAVA_OPTS, and $GRADLE_OPTS can contain fragments of
-#     shell script including quotes and variable substitutions, so put them in
-#     double quotes to make sure that they get re-expanded; and
-#   * put everything else in single quotes, so that it's not re-expanded.
+# Collect all arguments for the java command:
+#   * DEFAULT_JVM_OPTS, JAVA_OPTS, JAVA_OPTS, and optsEnvironmentVar are not allowed to contain shell fragments,
```
Comment (az108) [inline src/main/resources/templates/java/plain_gradle/exercise/gradlew.file:206]: JAVA_OPTS is mentioned 2 times?

**Review by az108 — COMMENTED** [review]
Rest works as intended and tested on TS1

File: src/main/resources/templates/java/plain_gradle/exercise/gradlew.file (hunk 1 above)
Comment (Strohgelaender) [reply src/main/resources/templates/java/plain_gradle/exercise/gradlew.file:206]: This is automatically generated by gradle.

**Review by pzdr7 — DISMISSED** [review]
Changes look consistent 👍

**Review by Jan-Thurner — COMMENTED** [review]
Tested on TS2. When enabling SCA and using `Gradle` as the project type, the build always fails.
image

**Review by Jan-Thurner — APPROVED** [review]
Tested again on TS2. Now the SCA works. The Builds of template and solution also were successful.

**Review by sarpsahinalp — APPROVED** [review]
Tested on ts2. Template and solution builds were successful. Also SCA feedback is shown.
image

**Review by egekurt123 — APPROVED** [review]
Works as expected on TS2

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by pzdr7 — APPROVED** [review]
Code LGTM

**Review by krusche — APPROVED** [review]

**Pull request merged**